tock/hours/views.py

```diff
@@ -512,17 +512,52 @@ class ReportingPeriodListView(PermissionMixin, ListView):
     """ Currently the home view that lists the completed and missing time
     periods """
     context_object_name = "incomplete_reporting_periods"
-    queryset = ReportingPeriod.objects.all()
+    queryset = ReportingPeriod.objects.all().order_by('-start_date')
     template_name = "hours/reporting_period_list.html"
     permission_classes = (IsAuthenticated, )
 
+    def disallowed_dates(self, date):
+        # If the end of the fiscal year is not on a weekend, then create
+        # a list of dates around (the buffer) the end of the fiscal year
+        # where a reporting period will not be automatically created. This
+        # prevents the automatic creation of a reporting period that has
+        # working week days that span two fiscal years.
+        fy_start_date = datetime.date(year=date.year, month=10, day=1)
+        if fy_start_date.weekday() < 5 and date.month < 10:
+            buffer_days = 7 # A week before and after.
+            buffer_start = fy_start_date - datetime.timedelta(days=buffer_days)
+            buffer_end = fy_start_date + datetime.timedelta(days=buffer_days)
+            return [ buffer_start + datetime.timedelta(days=i) for i in \
+                range((buffer_end - buffer_start).days) ]
+        else:
+            return []
+
+    def auto_create_reporting_period(self):
+        # Automatically creates a new reporting period if the latest reporting
+        # period has concluded.
+        latest_rp = self.queryset.first()
+        if not latest_rp:
+            return None # In case there are no reporting periods created yet.
+        if latest_rp.end_date >= datetime.datetime.utcnow().date():
+            return None # If the latest rp hasn't ended yet.
+        start_date = latest_rp.end_date + datetime.timedelta(days=1)
+        if start_date not in self.disallowed_dates(latest_rp.end_date):
+            ReportingPeriod.objects.create(
+                start_date=start_date,
+                end_date=start_date + datetime.timedelta(days=7),
+                max_working_hours=40,
+                min_working_hours=40,
+                exact_working_hours=40
+            )
+
     def get_context_data(self, **kwargs):
         context = super(
             ReportingPeriodListView, self).get_context_data(**kwargs)
+        self.auto_create_reporting_period()
```

Comment:
i would perhaps rename this function for clarity since most of the time it is not creating a new reporting period; maybe adding a 'maybe' would work :)

btw, if you're curious, here's how i would write this (note that it's nearly identical to yours and definitely doesn't need to be amended)

```
def maybe_create_new_reporting_period(self):
    latest_rp = self.queryset.first()

    if not latest_rp:
        return None

    latest_end = latest_rp.end_date
    new_start = latest_end + datetime.timedelta(days=1)
    today = datetime.datetime.utcnow().date()

    if latest_end < today and new_start not in self.disallowed_dates(latest_end):
        ReportingPeriod.objects.create(
            start_date=new_start,
            end_date=new_start + datetime.timedelta(days=7),
            max_working_hours=40,
            min_working_hours=40,
            exact_working_hours=40
        )
```